starlite/datastructures/provide.py

```diff
@@ -68,3 +83,114 @@ def __eq__(self, other: Any) -> bool:
             and other.use_cache == self.use_cache
             and other.value == self.value
         )
+
+
+class DependencyCleanupGroup:
+    """Wrapper for generator based dependencies.
+
+    Simplify cleanup by wrapping `next`/`anext` calls in `BackgroundTasks` and providing facilities to `throw` /
+    `athrow` into all generators consecutively. An instance of this class can be used as a contextmanager, which will
+    automatically throw any exceptions into its generators. All exceptions caught in this manner will be re-raised after
+    they have been thrown in the generators.
+    """
+
+    __slots__ = ("_generators", "_closed")
+
+    def __init__(self, generators: Optional[List["AnyGenerator"]] = None) -> None:
+        """Initialize `DependencyCleanupGroup`.
+
+        Args:
+            generators: An optional list of generators to be called at cleanup
+        """
+        self._generators = generators or []
+        self._closed = False
+
+    def add(self, generator: Union[Generator[Any, None, None], AsyncGenerator[Any, None]]) -> None:
+        """Add a new generator to the group.
+
+        Args:
+            generator: The generator to add
+
+        Returns:
+            None
+        """
+        if self._closed:
+            raise RuntimeError("Cannot call cleanup on a closed DependencyCleanupGroup")
+        self._generators.append(generator)
+
+    def _wrap_next(self, generator: "AnyGenerator") -> Callable[[], Coroutine[None, None, None]]:
+        if isasyncgen(generator):
```

Comment:
better to pre-calculate this in the init function and ensure that everything is an async callable.

Reply:
That won't work. The generators are not known at initialization time of the class, since they're being dynamically added with the `add` method.